File: lib.py/parser.py
from tokenizer import XMLTokenizer

import treebuilders
from constants import spaceCharacters
from treebuilders import simpletree

class XMLParser(object):
    def __init__(self, tree=simpletree.TreeBuilder):
        self.tree = tree()
        self.errors = []

        self.phases = {
            "start":self.startPhase,
            "main":self.mainPhase,
            "end":self.endPhase
        }
        self.phase = self.phases["start"]

    def _parse(self, stream, encoding=None):
        self.tree.reset()
        self.errors = []
        self.tokenizer = XMLTokenizer(stream, encoding)

        for token in self.tokenizer:
            self.phase(token)
        # When the loop finishes it's EOF
        # XXX

    def parse(self, stream, encoding=None):
        self._parse(stream, encoding=encoding)
        return self.tree.getDocument()

    def containsWhiteSpace(self, string):
        for c in string:
            if c not in spaceCharacters:
                return False
        return True

    def startPhase(self, token):
        if token["type"] == "StartTag":
            element = self.tree.createElement(token["name"], token["attributes"])
            self.tree.document.appendChild(element)
            self.tree.openElements.append(element)
            self.phase = self.phases["main"]
        elif token["type"] == "EmptyTag":
            element = self.tree.createElement(token["name"], token["attributes"])
            self.tree.document.appendChild(element)
            self.phase = self.phases["end"]
        elif token["type"] == "Comment":
            self.tree.document.appendChild(self.tree.commentClass(token["data"]))
        elif token["type"] == "Pi":
            self.tree.document.appendChild(self.tree.piClass(token["name"], token["data"]))
        elif token["type"] == "Characters" and self.containsWhiteSpace(token["data"]):
            pass
        else:
            # XXX parse error
            pass

    def mainPhase(self, token):
        if token["type"] == "Characters":
            self.tree.insertText(token["data"])
        elif token["type"] == "StartTag":
            element = self.tree.createElement(token["name"], token["attributes"])
            self.tree.openElements[-1].appendChild(element)
            self.tree.openElements.append(element)
        elif token["type"] == "EmptyTag":
            element = self.tree.createElement(token["name"], token["attributes"])
            self.tree.openElements[-1].appendChild(element)
        elif token["type"] == "EndTag":
            if self.tree.openElements[-1].name != token["name"]:
                # XXX parse error
                pass
            if self.tree.elementInScope(token["name"]):
                while self.tree.openElements[-1].name != token["name"]:
                    self.tree.openElements.pop()
                self.tree.openElements.pop()
                if len(self.tree.openElements) == 0:
                    self.phase = self.phases["end"]
        elif token["type"] == "EndTagShort":
            self.tree.openElements.pop()
            if len(self.tree.openElements) == 0:
                self.phase = self.phases["end"]
        elif token["type"] == "Comment":
            self.tree.openElements[-1].appendChild(self.tree.commentClass(token["data"]))
        elif token["type"] == "Pi":
            self.tree.openElements[-1].appendChild(self.tree.piClass(token["name"], token["data"]))

    def endPhase(self, token):
        if token["type"] == "Comment":
            self.tree.document.appendChild(self.tree.commentClass(token["data"]))
        elif token["type"] == "Pi":
            self.tree.document.appendChild(self.tree.piClass(token["name"], token["data"]))
        elif token["type"] == "Characters" and self.containsWhiteSpace(token["data"]):
            pass
        else:
            # XXX parse error
            pass

File: lib.rb/parser.rb
require 'tokenizer'
require 'treebuilders'
require 'constants'

class XMLParser
  def initialize(tree=SimpleTree::TreeBuilder)
    @tree = tree.new()
    @errors = []

    @phases = {
      "start" => :startPhase,
      "main" => :mainPhase,
      "end" => :endPhase
    }
    @phase = @phases["start"]
  end

  def _parse(stream, encoding=nil)
    @tree.reset()
    @errors = []
    @tokenizer = XMLTokenizer.new(stream, encoding)

    @tokenizer.each do |token|
      send @phase, token
    end
    # When the loop finishes it's EOF
    # XXX
  end

  def parse(stream, encoding=nil)
    _parse(stream, encoding=encoding)
    return @tree.getDocument()
  end

  def containsWhiteSpace(string)
    string.each_char do |c|
      if SpaceCharacters.include? c
        return false
      end
    end
    return true
  end

  def startPhase(token)
    if token["type"] == "StartTag"
      element = @tree.createElement(token["name"], token["attributes"])
      @tree.document.appendChild(element)
      @tree.openElements << element
      @phase = @phases["main"]
    elsif token["type"] == "EmptyTag"
      element = @tree.createElement(token["name"], token["attributes"])
      @tree.document.appendChild(element)
      @phase = @phases["end"]
    elsif token["type"] == "Comment"
      @tree.document.appendChild(@tree.commentClass.new(token["data"]))
    elsif token["type"] == "Pi"
      @tree.document.appendChild(@tree.piClass(token["name"], token["data"]))
    elsif token["type"] == "Characters" and containsWhiteSpace(token["data"])
      # pass
    else
      # XXX parse error
      # pass
    end
  end

  def mainPhase(token)
    if token["type"] == "Characters"
      @tree.insertText(token["data"])
    elsif token["type"] == "StartTag"
      element = @tree.createElement(token["name"], token["attributes"])
      @tree.openElements[-1].appendChild(element)
      @tree.openElements << element
    elsif token["type"] == "EmptyTag"
      element = @tree.createElement(token["name"], token["attributes"])
      @tree.openElements[-1].appendChild(element)
    elsif token["type"] == "EndTag"
      if @tree.openElements[-1].name != token["name"]
        # XXX parse error
        # pass
      end
      if @tree.elementInScope(token["name"])
        while @tree.openElements[-1].name != token["name"]
          @tree.openElements.pop()
        end
        @tree.openElements.pop()
        if @tree.openElements.empty?
          @phase = @phases["end"]
        end
      end
    elsif token["type"] == "EndTagShort"
      @tree.openElements.pop()
      if @tree.openElements.empty?
        @phase = @phases["end"]
      end
    elsif token["type"] == "Comment"
      @tree.openElements[-1].appendChild(@tree.commentClass(token["data"]))
    elsif token["type"] == "Pi"
      @tree.openElements[-1].appendChild(@tree.piClass(token["name"], token["data"]))
    end
  end

  def endPhase(token)
    if token["type"] == "Comment"
      @tree.document.appendChild(@tree.commentClass(token["data"]))
    elsif token["type"] == "Pi"
      @tree.document.appendChild(@tree.piClass(token["name"], token["data"]))
    elsif token["type"] == "Characters" and containsWhiteSpace(token["data"])
      # pass
    else
      # XXX parse error
      # pass
    end
  end
end
